### ISSUE TITLE
Log format: allow logging without source, or shortened source

### ISSUE THREAD

**Issue by howardjohn** [envoyproxy/envoy#10688]
*Title*: Log format: allow logging without source, or shortened source

*Description*:
Currently our logs look like `[bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:92] some log`. Due to covid-19 WFH my monitor is really tiny so now the source part `grpc_stream.h:92` takes up the entire screen.

It would be nice if there was some way to make this shorter. I looked over https://www.envoyproxy.io/docs/envoy/latest/operations/cli#cmdoption-log-format but did not see any, just `%v` which includes both the message and the source.

Just the tail part of the source might be a useful compromise as well (eg `[grpc_stream.h:92] some log`. If we wanted maximum flexibility, it seems like we could have

* `%v` full source + message
* `%something` full source
* `%something` trimmed source
* `%something` just the message

I am not sure that level is actually needed though

**Comment by mattklein123** [comment]
I would very much love this feature. I thought about implementing this a while ago, but unfortunately it's not very easy as the log file/line come from a macro and do not actually get passed through to spdlog where format strings are parsed. I think this could be fixed by making the file/line part of the logging function call itself and handling somehow, but it will be a little tricky.

cc @jmarantz @ggreenway

**Comment by ggreenway** [comment]
Yeah, currently the preprocessor builds the string `[filename.c:1234]`, and then does preprocessor/compile-time concatenation with the format string of the log message. We would either need some amazing preprocessor-fu to strip the prefix off the filename string, or somehow break apart the _VA_ARGS, prefix an extra `"{}:{}"` to the front of the format string using preprocessor/compile-time concat, strip the prefix from filename, pass two additional args after the format string, then pass the remaining _VA_ARGS_. It could be done, but it would take some effort.

**Comment by jmarantz** [comment]
crazy idea number 59291: you could also post-process the formatted text to clean up the filenames (ducks).

**Comment by lambdai** [comment]
It would be also helpful if we increase a little bit cardinality in file name and class name  
We have at list 16 classes named `Config` and 78 files named `config.cc`

**Comment by jmarantz** [comment]
Class names are not involved here, right? I think maybe we could just smart-strip the leading junk from the filename and get `[common/config/grpc_stream.h:92]` -- would that be good enough?

**Comment by mattklein123** [comment]
> crazy idea number 59291: you could also post-process the formatted text to clean up the filenames (ducks).

Honestly, this is a hack that I would seriously consider. See this tool that I wrote a long time ago to "fix" bazel paths in my IDE: https://github.com/envoyproxy/envoy/blob/master/tools/path_fix.sh. You could slightly modify that and then just pipe your output through that or run it using that script.

**Comment by euroelessar** [comment]
Based on [spdlog documentation](https://github.com/gabime/spdlog/wiki/3.-Custom-formatting) (and codebase) it supports file names/lines/function names in the log.
Envoy just needs to pass `spdlog::source_loc` as one of the arguments to [`log` method](https://github.com/gabime/spdlog/blob/ec12770693edc30988052718a6ec218b936b8f50/include/spdlog/logger.h#L219).

**Comment by euroelessar** [comment]
Getting to the next step (stripping bazel junk) can be done using [spdlog::custom_flag_formatter](https://github.com/gabime/spdlog/wiki/3.-Custom-formatting#extending-spdlog-with-your-own-flags) using c++ api of spdlog. It has access to `source_loc` and due to being just c++ code can be easily tested.

**Comment by mattklein123** [comment]
@euroelessar nice find! Awesome!
